warp_drive/training/lightning-trainer.py

```diff
@@ -0,0 +1,520 @@
+
+            # Initialize episodic rewards and push to the GPU
+            num_agents_for_policy = len(self.policy_tag_to_agent_id_map[policy])
+            self.num_completed_episodes[policy] = 0
+            self.episodic_reward_sum[policy] = (
+                torch.tensor(0).type(torch.float32).cuda()
+            )
+            self.reward_running_sum[policy] = torch.zeros(
+                (self.num_envs, num_agents_for_policy)
+            ).cuda()
+
+        # Initialize the trainers
+        self.trainers = {}
+        self.clip_grad_norm = {}
+        self.max_grad_norm = {}
+        for policy in self.policies_to_train:
+            algorithm = self._get_config(["policy", policy, "algorithm"])
+            assert algorithm in ["A2C", "PPO"]
+            entropy_coeff = self._get_config(["policy", policy, "entropy_coeff"])
+            vf_loss_coeff = self._get_config(["policy", policy, "vf_loss_coeff"])
+            self.clip_grad_norm[policy] = self._get_config(
+                ["policy", policy, "clip_grad_norm"]
+            )
+            if self.clip_grad_norm[policy]:
+                self.max_grad_norm[policy] = self._get_config(
+                    ["policy", policy, "max_grad_norm"]
+                )
+            normalize_advantage = self._get_config(
+                ["policy", policy, "normalize_advantage"]
+            )
+            normalize_return = self._get_config(["policy", policy, "normalize_return"])
+            gamma = policy_config["gamma"]
+
+            if algorithm == "PPO":
+                clip_param = self._get_config(["policy", policy, "clip_param"])
+
+            if algorithm == "A2C":
+                # Advantage Actor-Critic
+                self.trainers[policy] = A2C(
+                    discount_factor_gamma=gamma,
+                    normalize_advantage=normalize_advantage,
+                    normalize_return=normalize_return,
+                    vf_loss_coeff=vf_loss_coeff,
+                    entropy_coeff=entropy_coeff,
+                )
+                logging.info(f"Initializing the A2C trainer for policy {policy}")
+            elif algorithm == "PPO":
+                # Proximal Policy Optimization
+                self.trainers[policy] = PPO(
+                    discount_factor_gamma=gamma,
+                    clip_param=clip_param,
+                    normalize_advantage=normalize_advantage,
+                    normalize_return=normalize_return,
+                    vf_loss_coeff=vf_loss_coeff,
+                    entropy_coeff=entropy_coeff,
+                )
+                logging.info(f"Initializing the PPO trainer for policy {policy}")
+            else:
+                raise NotImplementedError
+
+        # Push all the data and tensor arrays to the GPU
+        # upon resetting environments for the very first time.
+        self.cuda_envs.reset_all_envs()
+
+        # Create and push data placeholders to the device
+        create_and_push_data_placeholders(
```

Comment:
This doesn't seem to be at the right place. This seems related to the data instead.

Reply:
Here is another abstraction used with Lightning for RL: https://github.com/lebrice/Sequoia/blob/master/examples/basic/pl_example.py

Reply:
This is just a one-time operation for setting aside memory on the GPU. I could move it to a DataModule, but seemed unnecessary.